DOCSP-30530: enterprise auth (#29)
mongodb/docs-rust: source/fundamentals/enterprise-auth.txt

.. _rust-enterprise-auth:

====================================
Enterprise Authentication Mechanisms
====================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to authenticate to MongoDB using the
**authentication mechanisms** available in the MongoDB Enterprise

> Edition. When you connect to MongoDB, you can use an authentication

Reviewer: S: Consider removing "you ~can~" throughout and use imperative instead.

Author: Will try to implement where it sounds natural

mechanism to establish trust between the driver and the server.

> The driver supports only the :ref:`LDAP (Plain) <rust-LDAP>`

Reviewer: ```suggestion
The {+driver-short+} supports only the :ref:`LDAP (PLAIN) <rust-LDAP>`
```

> authentication mechanism. To learn how to use other methods to

Reviewer: S: Make a new graf for Kerberos, and mention that it is not supported by the driver.

authenticate by using the GSSAPI/Kerberos authentication mechanism, see
:manual:`Kerberos Authentication </core/kerberos/>` in the Server manual.

.. tip::

   To authenticate to MongoDB by using mechanisms available in
   the MongoDB Community Edition, see the guide on :ref:`<rust-authentication>`.

   To learn more about connecting to a MongoDB deployment, see the
   :ref:`rust-connect-to-mongodb`.

To select a specific authentication mechanism, you can specify the
mechanism, your credentials, and other necessary information
in the options of your connection string or in a ``Credential`` struct.

.. _rust-LDAP:

Authenticate to LDAP (PLAIN)
----------------------------

You can authenticate to a Lightweight Directory Access Protocol (LDAP) server 
by using your directory server username and password.

.. warning::

>    This authentication mechanism sends the password to the server in

Reviewer: ```suggestion
   This authentication mechanism sends your password to the server in
```

>    plaintext, so you should use this mechanism only after enabling TLS

Reviewer: ```suggestion
   plaintext. Only use this mechanism only after enabling TLS
```

   on your connection.
   
   To learn more, see :manual:`TLS/SSL (Transport Encryption)
   </core/security-transport-encryption/>` in the Server manual.

Example
~~~~~~~

To specify the ``PLAIN`` authentication mechanism, set the
``mechanism`` field of your ``Credential`` struct to

> ``AuthMechanism::Plain``. This example specifies the

Reviewer: ```suggestion
``AuthMechanism::Plain`` and pass it to client_options.builder. This example specifies the
```

authentication mechanism by using the following placeholders:

- ``username``: Your LDAP username
- ``password``: Your LDAP password

.. literalinclude:: /includes/fundamentals/code-snippets/enterprise-auth.rs
   :language: rust
   :dedent:
   :start-after: start-ldap
   :end-before: end-ldap

.. note:

   Since your credentials are stored outside of MongoDB, you must use the
   ``$external`` database for authentication. The ``source`` field of

>    the ``Credential`` struct defaults to this database name, so you do

Reviewer: ```suggestion
   the ``Credential`` struct defaults to ``$external``, so you do
```

   not need to explicitly set this field.

> You can alternatively authenticate by using a connection string URI by

Reviewer: ```suggestion
Alternatively, you can authenticate by using a connection string URI by
```

setting the value of the ``authMechanism`` connection string option to ``PLAIN``.

> This example shows how to specify this authentication mechanism in

Reviewer: ```suggestion
This example shows how to specify the authentication mechanism in
```

a connection string URI by using the following placeholders:

- ``username``: Your LDAP username
- ``password``: Your LDAP password
- ``hostname``: The network address of your MongoDB server

.. code-block:: rust
   
   let uri = "mongodb://<username>:<password>@<hostname>/?authMechanism=PLAIN";

.. note::

   The authentication mechanism refers to PLAIN instead of LDAP since it

>    authenticates using the PLAIN Simple Authentication and Security Layer

Reviewer: S: Consider moving up. And maybe something more clear, like: "The LDAP mech is called PLAIN because..."

   (SASL) defined in `RFC-4616 <https://tools.ietf.org/html/rfc4616>`__.

Additional Information
----------------------

To learn more about the concepts in this guide, see the following documentation:

- :manual:`MongoDB Server Support for LDAP Proxy Authentication </core/security-ldap/>`

..
  - :ref:`rust-connection-options`

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about the methods and types discussed in this

> guide, see the following API Documentation:

Reviewer: ```suggestion
guide, see the following Rust API Documentation:
```

Author: seems self-explanatory as is, plus sometimes this section links to other documentation than the driver, such as in [this PR](https://github.com/mongodb/docs-rust/pull/28/files)

mechanism to establish trust between the driver and the server.

The {+driver-short+} supports only the :ref:`LDAP (PLAIN) <rust-LDAP>`
authentication mechanism.

.. note:: GSSAPI/Kerberos Authentication
   
   The driver does not support the GSSAPI/Kerberos authentication
   mechanism, but you can use other methods to authenticate this way. To
   learn more about these methods, see :manual:`Kerberos Authentication
   </core/kerberos/>` in the Server manual.

.. seealso::

   To authenticate to MongoDB by using mechanisms available in
   the MongoDB Community Edition, see the guide on :ref:`<rust-authentication>`.

   To learn more about connecting to a MongoDB deployment, see the
   :ref:`rust-connect-to-mongodb`.

To select a specific authentication mechanism, specify the
mechanism, your credentials, and other necessary information
in the options of your connection string or in a ``Credential`` struct.

.. _rust-LDAP:

Authenticate to LDAP (PLAIN)
----------------------------

You can authenticate to a Lightweight Directory Access Protocol (LDAP) server 
by using your directory server username and password.

The name of the authentication mechanism is ``PLAIN`` instead of LDAP

> because the mechanism authenticates by using the PLAIN Simple Authentication and

Reviewer: ```suggestion
because the mechanism uses the PLAIN Simple Authentication and
```

   :language: rust
   :dedent:
   :start-after: start-ldap
   :end-before: end-ldap

.. note:

   Since your credentials are stored outside of MongoDB, you must use the
   ``$external`` database for authentication. The ``source`` field of
   the ``Credential`` struct defaults to ``$external``, so you do
   not need to explicitly set this field.

Alternatively, you can authenticate by using a connection string URI by
setting the value of the ``authMechanism`` connection string option to ``PLAIN``.
This example shows how to specify the ``PLAIN`` authentication mechanism in
a connection string URI by using the following placeholders:

- ``username``: Your LDAP username
- ``password``: Your LDAP password
- ``hostname``: The network address of your MongoDB server

.. code-block:: rust
   
   let uri = "mongodb://<username>:<password>@<hostname>/?authMechanism=PLAIN";

Additional Information
----------------------

To learn more about the concepts in this guide, see the following documentation:

- :manual:`MongoDB Server Support for LDAP Proxy Authentication </core/security-ldap/>`
- :ref:`rust-connection-options`

> 

Reviewer: S: Maybe a link to https://www.mongodb.com/docs/manual/reference/connection-string/ .

.. _rust-enterprise-auth:

====================================
Enterprise Authentication Mechanisms
====================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to authenticate to MongoDB using the
**authentication mechanisms** available in the MongoDB Enterprise
Edition. When you connect to MongoDB, you can use an authentication
mechanism to establish trust between the driver and the server.

> The {+driver-short+} supports only the :ref:`LDAP (PLAIN) <rust-LDAP>`
> authentication mechanism.

Reviewer 2: ```suggestion
The {+driver-short+} supports the :ref:`LDAP (PLAIN) <rust-LDAP>`
enterprise authentication mechanism.
```

Suggest removing "only" and clarifying this applies to "enterprise". The non-enterprise authentication mechanisms are an available alternative (SCRAM + X509).

Example
~~~~~~~

To specify the ``PLAIN`` authentication mechanism, set the
``mechanism`` field of your ``Credential`` struct to
``AuthMechanism::Plain``. This example specifies the
authentication mechanism by using the following placeholders:

- ``username``: Your LDAP username
- ``password``: Your LDAP password

.. literalinclude:: /includes/fundamentals/code-snippets/enterprise-auth.rs
   :language: rust
   :dedent:
   :start-after: start-ldap
   :end-before: end-ldap

.. note:

   Since your credentials are stored outside of MongoDB, you must use the
   ``$external`` database for authentication. The ``source`` field of
   the ``Credential`` struct defaults to ``$external``, so you do
   not need to explicitly set this field.

Alternatively, you can authenticate by using a connection string URI by
setting the value of the ``authMechanism`` connection string option to ``PLAIN``.
This example shows how to specify the ``PLAIN`` authentication mechanism in
a connection string URI by using the following placeholders:

- ``username``: Your LDAP username
- ``password``: Your LDAP password
- ``hostname``: The network address of your MongoDB server

.. code-block:: rust
   

>    let uri = "mongodb://<username>:<password>@<hostname>/?authMechanism=PLAIN";

Reviewer 2: ```suggestion
   let uri = "mongodb://<username>:<password>@<hostname>/?authSource=$external&authMechanism=PLAIN";
```

Use of `$external` is shown in the Java docs: https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/enterprise-auth/
